### PR TITLE
types/json: add binary json functions

### PR DESCRIPTION
**Pull request by coocood**
All the JSON functions have been supported by BinaryJSON.

### CODE REVIEW

**Comment by coocood** [comment]
@hicqu @jackysp PTAL

File: types/json/binary_test.go [hunk 1]
```diff
@@ -46,11 +46,275 @@ func (s *testJSONSuite) TestBinaryJSONMarshalUnmarshal(c *C) {
...
-		msg := fmt.Sprintf("ParseFromString(%s) fail", s)
-		panic(msg)
+	bj, err := ParseBinaryFromString(s)
+	c.Check(err, IsNil)
+	return bj
+}
+
+const benchStr = `{"a":[1,"2",{"aa":"bb"},4,null],"b":true,"c":null}`
+
+func BenchmarkBinaryMarshal(b *testing.B) {
```
Comment (jackysp) [inline types/json/binary_test.go]: Any results for these?

File: types/json/binary.go [hunk 2]
```diff
@@ -202,6 +205,53 @@ func (bj BinaryJSON) getFloat64() float64 {
...
+func (bj BinaryJSON) getString() []byte {
+	strLen, lenLen := uint64(bj.Value[0]), 1
+	if strLen >= 0x80 {
+		strLen, lenLen = binary.Uvarint(bj.Value)
+	}
+	return bj.Value[lenLen : lenLen+int(strLen)]
+}
+
+func (bj BinaryJSON) getElemCount() int {
+	return int(binary.LittleEndian.Uint32(bj.Value))
```
Comment (jackysp) [inline types/json/binary.go]: Why use `LittleEndian` here instead of use `endian`?

File: types/json/binary_test.go (hunk 1 above)
Comment (coocood) [reply types/json/binary_test.go]: 15x faster

File: types/json/binary.go [hunk 3]
```diff
@@ -202,6 +205,53 @@ func (bj BinaryJSON) getFloat64() float64 {
...
+}
+
+func (bj BinaryJSON) arrayGetElem(idx int) BinaryJSON {
+	return bj.valEntryGet(8 + idx*5)
+}
+
+func (bj BinaryJSON) objectGetKey(i int) []byte {
```
Comment (jackysp) [inline types/json/binary.go]: So many magic numbers in this part.

**Comment by coocood** [comment]
@jackysp PTAL

File: types/json/binary.go [hunk 4]
```diff
@@ -148,6 +148,15 @@ var (
 	endian   = binary.LittleEndian
 )
 
+const (
```
Comment (jackysp) [inline types/json/binary.go]: I think we can add a comment here. e.g.
```
// -------------------------------------------------------------------------------
// | header                        | key                                    | value
// -------------------------------------------------------------------------------
// | element size | data size | key off | key length | val type
// -------------------------------------------------------------------------------
// | 8 bytes length (little endian) |  x bytes | 1 byte
// -------------------------------------------------------------------------------
```

Comment (coocood) [reply types/json/binary.go]: @jackysp 
The full layout of binary json can be found at `serdes.go`, we can move it to `binary.go` when we delete the `serdes.go`.

File: types/json/binary.go [hunk 5]
```diff
@@ -202,6 +214,53 @@ func (bj BinaryJSON) getFloat64() float64 {
 	return math.Float64frombits(bj.getUint64())
 }
 
+func (bj BinaryJSON) getString() []byte {
+	strLen, lenLen := uint64(bj.Value[0]), 1
+	if strLen >= 0x80 {
```
Comment (jackysp) [inline types/json/binary.go]: 0x80 should be utf8.RuneSelf ?

File: types/json/binary_functions.go [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+	switch bj.TypeCode {
+	case TypeCodeString:
+		s, err := unquoteString(hack.String(bj.getString()))
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		// Remove prefix and suffix '"' and '\''.
+		slen := len(s)
+		if slen > 0 {
+			head, tail := s[0], s[slen-1]
```
Comment (jackysp) [inline types/json/binary_functions.go]: if s := " or ', what will happen?

File: types/json/binary.go (hunk 4 above)
Comment (jackysp) [reply types/json/binary.go]: ok

**Comment by shenli** [comment]
@hicqu PTAL

File: types/json/binary_functions.go [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+
+// Unquote is for JSON_UNQUOTE.
+func (bj BinaryJSON) Unquote() (string, error) {
+	switch bj.TypeCode {
+	case TypeCodeString:
+		s, err := unquoteString(hack.String(bj.getString()))
+		if err != nil {
+			return "", errors.Trace(err)
+		}
+		// Remove prefix and suffix '"' and '\''.
```
Comment (hicqu) [inline types/json/binary_functions.go]: We don't need to remove '\''. It's a bug. Maybe we can fix it in next PR.

**Comment by hicqu** [comment]
LGTM.

**Pull request merged**